### DISCUSSION TITLE
Handling concurrent requests with different database sessions in Django using SQLAlchemy

### DISCUSSION THREAD

**Discussion by akashchevli** [sqlalchemy/sqlalchemy#11051] [Usage Questions]
I have a Django application where each brand has its own database, and incoming requests specify the brand ID. I'm using SQLAlchemy for database access.

Here's the flow:

1. The client sends a request with the brand ID to Django.
2. Django middleware fetches the session object based on the brand ID.
3. I'm using a function (**replace_session**) to replace the session object with the incoming brand request session (**replace_session** function calling from the Middleware).

This works fine when requests arrive one after the other. However, I'm concerned about handling concurrent requests, especially when multiple requests arrive at the same time.

Below is my code,

```
engine = create_engine(<Database URL>)
session = scoped_session(sessionmaker(bind=engine, autocommit=True))
Base.metadata.create_all(engine)
```

```
brand_a = create_engine(<Database URL>)
brand_a_session = scoped_session(sessionmaker(bind=brand_a, autocommit=True))
Base.metadata.create_all(brand_a)
```

```
brand_b = create_engine(<Database URL>)
brand_b_session = scoped_session(sessionmaker(bind=brand_b, autocommit=True))
Base.metadata.create_all(brand_b)
```

```
project_session = None
def replace_session(project_session):
    project_session = session
    return project_session
```

**Q. Why am I passing the session object to the project_session variable?**
**A**. The reason is that, currently, I have only one database, and I'm using **session** variables to write the query and fetch the data from the database. Now, I've got multiple databases, and I don't want to replace the **session** variable with the requested session object(**project_session**) because there are lots of methods that use the **session** variable to fetch the records. This is the only reason to pass the session value to the **project_session** variable.

Note: **replace_session** is calling from the Middleware with brand session object.

Consider the following scenario:

Request 1:

- Brand ID: A
- Middleware fetches the session object for brand A.
- **replace_session** function replaces the session object with the session for brand A.

Request 2:

- Brand ID: B
- Middleware fetches the session object for brand B.
- **replace_session** function replaces the session object with the session for brand B.

If both requests arrive simultaneously, will each request have its own session object, or could there be interference between them?

I want to ensure that each request is processed independently with its own database session and that changes made to one session do not affect another session.

Is this approach safe for handling concurrent requests with different database sessions in Django using SQLAlchemy? If not, what potential issues could arise, and how can I address them?

**Comment by CaselIT** [comment]
Hi,

Not sure what you are doing with replace session, but that's not really safe even if you had only a single database, since you could still have concurrent requests and that would be changing the session mid request.

I assume that django, like almost all web framework, has a way of setting "request-local" variables, either in a request object dict or using thread-local variables or similar.
You should use that feature to associate a particular session with an http request of django, so that multiple request would not interfere with each other. That should probably be done by the middleware so that the request handing code already has the session setup.

I assume how do do this will be well documented on the django website
